cli/command/stack/list.go

```diff
@@ -18,20 +22,91 @@ func newListCommand(dockerCli command.Cli) *cobra.Command {
-			if dockerCli.ClientInfo().HasKubernetes() {
-				kli, err := kubernetes.WrapCli(dockerCli, kubernetes.NewOptions(cmd.Flags()))
-				if err != nil {
-					return err
-				}
-				return kubernetes.RunList(kli, opts)
-			}
-			return swarm.RunList(dockerCli, opts)
+			return runList(cmd, dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
 	flags.StringVar(&opts.Format, "format", "", "Pretty-print stacks using a Go template")
+	flags.StringSliceVar(&opts.Namespaces, "namespace", []string{}, "Kubernetes namespaces to use")
+	flags.SetAnnotation("namespace", "kubernetes", nil)
+	flags.SetAnnotation("namespace", "experimentalCLI", nil)
 	flags.BoolVarP(&opts.AllNamespaces, "all-namespaces", "", false, "List stacks among all Kubernetes namespaces")
 	flags.SetAnnotation("all-namespaces", "kubernetes", nil)
+	flags.SetAnnotation("all-namespaces", "experimentalCLI", nil)
 	return cmd
 }
+
+func runList(cmd *cobra.Command, dockerCli command.Cli, opts options.List) error {
+	stacks := []*formatter.Stack{}
+	if dockerCli.ClientInfo().HasSwarm() {
+		ss, err := swarm.GetStacks(dockerCli)
+		if err != nil {
+			return err
+		}
+		stacks = append(stacks, ss...)
+	}
+	if dockerCli.ClientInfo().HasKubernetes() {
+		kopts := kubernetes.NewOptions(cmd.Flags())
+		if opts.AllNamespaces || len(opts.Namespaces) == 0 {
+			if dockerCli.ClientInfo().HasAll() {
+				opts.AllNamespaces = true
+			}
+			ss, err := getStacks(dockerCli, opts, kopts)
+			if err != nil {
+				return err
+			}
+			stacks = append(stacks, ss...)
+		} else {
+			for _, nm := range removeDuplicates(opts.Namespaces) {
+				kopts.Namespace = nm
+				ss, err := getStacks(dockerCli, opts, kopts)
+				if err != nil {
+					return err
+				}
+				stacks = append(stacks, ss...)
+			}
+		}
+	}
+	format := opts.Format
+	if format == "" || format == formatter.TableFormatKey {
+		format = formatter.SwarmStackTableFormat
+		if dockerCli.ClientInfo().HasKubernetes() {
+			format = formatter.KubernetesStackTableFormat
+		}
+	}
+	stackCtx := formatter.Context{
+		Output: dockerCli.Out(),
+		Format: formatter.Format(format),
+	}
+	sort.Slice(stacks, func(i, j int) bool {
+		return sortorder.NaturalLess(stacks[i].Name, stacks[j].Name) ||
+			!sortorder.NaturalLess(stacks[j].Name, stacks[i].Name) &&
+				sortorder.NaturalLess(stacks[j].Namespace, stacks[i].Namespace)
+	})
+	return formatter.StackWrite(stackCtx, stacks)
+}
+
+func getStacks(dockerCli command.Cli, opts options.List, kopts kubernetes.Options) ([]*formatter.Stack, error) {
+	kli, err := kubernetes.WrapCli(dockerCli, kopts)
+	if err != nil {
+		return nil, err
+	}
+	ss, err := kubernetes.GetStacks(kli, opts)
+	if err != nil {
+		return nil, err
+	}
+	return ss, nil
+}
+
+func removeDuplicates(namespaces []string) []string {
+	mnms := map[string]struct{}{}
+	for _, nm := range namespaces {
+		mnms[nm] = struct{}{}
+	}
+	namespaces = []string{}
+	for nm := range mnms {
+		namespaces = append(namespaces, nm)
```

Comment:
no need to loop twice here, could be something like;

```go
func removeDuplicates(namespaces []string) []string {
	found := make(map[string]bool)
	results := namespaces[:0]
	for _, n := range namespaces {
		if !found[n] {
			results = append(results, n)
			found[n] = true
		}
	}
	return results
}
```

Reply:
Fair enough, I'll change that.